File: Golang/DynamicProgramming/unboundedKnapsack.go
package main

func unboundedKnapsack(N int, W int, val []int, wt []int) int {
	dp := make([][]int, N+1)
	for i := range dp {
		dp[i] = make([]int, W+1)
	}

	for i := 1; i <= N; i++ {
		for j := 1; j <= W; j++ {

			exclude := dp[i-1][j]

			include := 0
			if wt[i-1] <= j {
				include = val[i-1] + dp[i][j-wt[i-1]]
			}

			if include > exclude {
				dp[i][j] = include
			} else {
				dp[i][j] = exclude
			}
		}
	}

	return dp[N][W]
}

File: Python/DynamicProgramming/unboundedKnapsack.py
def unboundedKnapsack(N, W, val, wt):
    dp = [[0] * (W + 1) for _ in range(N + 1)]

    for i in range(1, N + 1):
        for j in range(1, W + 1):
            exclude = dp[i - 1][j]

            include = 0
            if wt[i - 1] <= j:
                include = val[i - 1] + dp[i][j - wt[i - 1]]

            dp[i][j] = max(include, exclude)

    return dp[N][W]